File: DittoDatto-old/services/mercury-engine/src/mercury_core/models/resource.py
"""Resource and ResourceGroup models.

Source: schemas/company-blueprint.surql §4.1–4.2
Zod origin: packages/shared-types/src/resource.ts

Resources represent physical or logical entities (rooms, tables, stations, equipment)
that can be assigned to bookings. The best-fit allocation algorithm in
resource-availability.ts uses these for optimal resource matching.
"""

from __future__ import annotations

from .common import (
    Currency,
    MercuryModel,
    ResourcePriority,
    ResourceType,
    TimestampMixin,
)


class ResourceGroup(MercuryModel, TimestampMixin):
    """Logical grouping of resources (e.g., 'Massage Rooms', 'Window Tables')."""

    id: str | None = None
    establishment: str = ""  # record<establishment>
    name: str = ""
    description: str | None = None
    sort_order: int = 0
    show_on_storefront: bool = False


class Resource(MercuryModel, TimestampMixin):
    """A bookable physical or logical resource.

    Used in the best-fit allocation algorithm during hold creation.
    """

    id: str | None = None
    establishment: str = ""  # record<establishment>
    resource_group: str | None = None  # record<resource_group>

    # Identity
    name: str = ""
    description: str | None = None
    type: ResourceType = ResourceType.ROOM

    # Capacity (table reservation mode)
    min_capacity: int = 1
    max_capacity: int = 1

    # Pricing (add-on resources)
    price: float | None = None
    currency: Currency = Currency.NOK

    image_url: str | None = None

    # Booking behavior
    is_bookable: bool = True
    allow_overlapping: bool = False
    booking_interval: int | None = None
    priority: ResourcePriority = ResourcePriority.NORMAL
    sort_order: int = 0

File: DittoDatto-old/packages/shared-types/src/resource.ts
import { z } from "zod";
import { IdSchema, DateTimeSchema } from "./common";

/**
 * Resource & Resource Group Schemas
 *
 * Resources are physical assets required to perform a service:
 * rooms, tables, chairs, stations, equipment, etc.
 *
 * A Service may require one or more Resources (via requiredResourceGroupIds).
 * MercuryEngine slot calculation checks:
 *   staff available AND resource available → slot is bookable.
 *
 * Firestore path: companies/{companyId}/stores/{storeId}/resources/{resourceId}
 * Firestore path: companies/{companyId}/stores/{storeId}/resourceGroups/{groupId}
 *
 * Inspired by Noona HQ API Resources — adapted for DittoDatto's multi-store model.
 * NOTE: Schema only. Engine wiring deferred until table reservation implementation.
 */

// ---------------------
// Resource Type
// ---------------------

export const ResourceTypeSchema = z.enum([
  "room",       // Massage room, treatment room, meeting room
  "table",      // Restaurant table, bar seating
  "station",    // Hair station, nail station, workbench
  "equipment",  // Projector, laser machine, specific tool
  "addon",      // Add-on item: Beer Keg, Bartender, Laser lights (has price)
]);
export type ResourceType = z.infer<typeof ResourceTypeSchema>;

export const ResourcePrioritySchema = z.enum(["low", "normal", "high"]);
export type ResourcePriority = z.infer<typeof ResourcePrioritySchema>;

// ---------------------
// Resource Group
// ---------------------

/**
 * Logical grouping of resources.
 * Example: "Massage Rooms" (contains Room 1, Room 2, Room 3)
 *
 * When a Service requires a resource, it references a ResourceGroup.
 * The engine picks any available resource from that group.
 */
export const ResourceGroupSchema = z.object({
  id: IdSchema,
  storeId: IdSchema,
  companyId: IdSchema,

  name: z.string().min(1).max(100),        // "Massage Rooms", "Window Tables"
  description: z.string().max(1000).optional(),

  // Display
  sortOrder: z.number().int().default(0),
  /** When true, resources in this group are browsable on the public store page (e.g. halls, studios) */
  showOnStorefront: z.boolean().default(false),

  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema,
});
export type ResourceGroup = z.infer<typeof ResourceGroupSchema>;

// ---------------------
// Resource
// ---------------------

/**
 * A single physical asset.
 *
 * Key design decisions:
 * - min/maxCapacity enables table reservation logic ("Table for 2–4")
 * - isBookable lets owners temporarily disable a resource (maintenance, etc.)
 * - resourceGroupId is optional: ungrouped resources are standalone
 */
export const ResourceSchema = z.object({
  id: IdSchema,
  storeId: IdSchema,
  companyId: IdSchema,
  resourceGroupId: IdSchema.optional(),

  name: z.string().min(1).max(100),        // "Massage Room 1", "Table 5", "Beer Keg"
  description: z.string().max(1000).optional(),
  type: ResourceTypeSchema,

  // Capacity (for table reservation mode)
  minCapacity: z.number().int().min(1).default(1),    // Min guests this resource serves
  maxCapacity: z.number().int().min(1).default(1),    // Max guests this resource serves

  // Pricing (for add-on resources: "Beer Keg 1500 kr")
  price: z.number().min(0).optional(),
  currency: z.string().length(3).default("NOK"),

  // Media
  imageUrl: z.string().url().optional(),

  // Booking behavior
  isBookable: z.boolean().default(true),
  allowOverlapping: z.boolean().default(false),  // Allow multiple bookings at same time
  bookingInterval: z.number().int().min(5).optional(), // Override store-level interval (minutes)
  priority: ResourcePrioritySchema.default("normal"),  // Auto-assignment preference

  // Display
  sortOrder: z.number().int().default(0),

  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema,
});
export type Resource = z.infer<typeof ResourceSchema>;

// ---------------------
// Input Schemas
// ---------------------

export const CreateResourceGroupSchema = ResourceGroupSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
});
export type CreateResourceGroup = z.infer<typeof CreateResourceGroupSchema>;

export const CreateResourceSchema = ResourceSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
});
export type CreateResource = z.infer<typeof CreateResourceSchema>;
